File: segfault-examples/02_use_after_free.cpp
#include <iostream>
#include <memory>

// Example 1: Classic use-after-free
void useAfterFree() {
    std::cout << "\n=== Use After Free ===" << std::endl;
    
    int* ptr = new int(42);
    std::cout << "Allocated: *ptr = " << *ptr << std::endl;
    
    delete ptr;
    std::cout << "Deleted ptr" << std::endl;
    
    // CRASH: Using freed memory
    std::cout << "Attempting to use freed memory..." << std::endl;
    *ptr = 100;
    std::cout << "Value after free: " << *ptr << std::endl;
}

// Example 2: Dangling pointer from stack
int* danglingStackPointer() {
    int local = 42;
    return &local;  // Returns address of destroyed stack variable
}

void useDanglingStack() {
    std::cout << "\n=== Dangling Stack Pointer ===" << std::endl;
    
    int* ptr = danglingStackPointer();
    std::cout << "Got pointer to destroyed stack variable" << std::endl;
    
    // CRASH: Accessing destroyed stack memory
    std::cout << "Attempting to use: " << *ptr << std::endl;
}

// Example 3: Reference to destroyed object
class MyClass {
public:
    int value;
    MyClass(int v) : value(v) {
        std::cout << "  Constructor: value = " << value << std::endl;
    }
    ~MyClass() {
        std::cout << "  Destructor: value was " << value << std::endl;
    }
};

MyClass& returnReference() {
    MyClass local(42);
    return local;  // Returns reference to object that's about to be destroyed
}

void useDanglingReference() {
    std::cout << "\n=== Dangling Reference ===" << std::endl;
    
    MyClass& ref = returnReference();
    std::cout << "Got reference to destroyed object" << std::endl;
    
    // CRASH: Object is destroyed
    std::cout << "Attempting to access: " << ref.value << std::endl;
}

// Example 4: Iterator invalidation
#include <vector>

void iteratorInvalidation() {
    std::cout << "\n=== Iterator Invalidation ===" << std::endl;
    
    std::vector<int> vec = {1, 2, 3, 4, 5};
    auto it = vec.begin();
    
    std::cout << "Iterator points to: " << *it << std::endl;
    
    // This invalidates all iterators
    vec.push_back(6);
    vec.push_back(7);
    vec.push_back(8);  // May cause reallocation
    
    // CRASH: Iterator is now invalid
    std::cout << "Attempting to use invalidated iterator..." << std::endl;
    std::cout << *it << std::endl;
}

// Example 5: Double use after free
void doubleUseAfterFree() {
    std::cout << "\n=== Multiple Uses After Free ===" << std::endl;
    
    int* ptr = new int(42);
    delete ptr;
    
    // CRASH: First use after free
    *ptr = 100;
    
    // CRASH: Second use after free
    std::cout << *ptr << std::endl;
}

// SAFE ALTERNATIVES
void safeAlternatives() {
    std::cout << "\n=== SAFE ALTERNATIVES ===" << std::endl;
    
    // 1. Smart pointers prevent use-after-free
    {
        std::cout << "\n1. Using unique_ptr:" << std::endl;
        std::unique_ptr<int> ptr = std::make_unique<int>(42);
        std::cout << "  Value: " << *ptr << std::endl;
        
        // After ptr is destroyed, you can't access it anymore
        // (compile-time safety with move semantics)
    }
    
    // 2. Shared ownership with shared_ptr
    {
        std::cout << "\n2. Using shared_ptr:" << std::endl;
        std::shared_ptr<int> ptr1 = std::make_shared<int>(42);
        {
            std::shared_ptr<int> ptr2 = ptr1;
            std::cout << "  ptr2 value: " << *ptr2 << std::endl;
        }  // ptr2 destroyed, but memory still valid
        std::cout << "  ptr1 still valid: " << *ptr1 << std::endl;
    }
    
    // 3. Use values instead of pointers when possible
    {
        std::cout << "\n3. Using values:" << std::endl;
        MyClass obj(100);
        std::cout << "  Direct value access: " << obj.value << std::endl;
        // Automatic cleanup, no pointers needed
    }
    
    // 4. RAII pattern
    {
        std::cout << "\n4. RAII with unique_ptr:" << std::endl;
        auto obj = std::make_unique<MyClass>(200);
        std::cout << "  Object value: " << obj->value << std::endl;
    }  // Automatic destruction
    
    std::cout << "\nAll memory safely managed!" << std::endl;
}

int main() {
    std::cout << "=== Use-After-Free Segfault Examples ===" << std::endl;
    std::cout << "Uncomment ONE example to see the crash\n" << std::endl;
    
    // UNCOMMENT ONE TO SEE CRASH:
    // useAfterFree();
    // useDanglingStack();
    // useDanglingReference();
    // iteratorInvalidation();
    // doubleUseAfterFree();
    
    // This is safe to run:
    safeAlternatives();
    
    std::cout << "\nProgram completed safely" << std::endl;
    
    return 0;
}

File: segfault-examples/02_use_after_free.rs
// Rust prevents use-after-free at compile time through ownership!

struct MyStruct {
    value: i32,
}

impl MyStruct {
    fn new(v: i32) -> Self {
        println!("  Constructor: value = {}", v);
        MyStruct { value: v }
    }
}

impl Drop for MyStruct {
    fn drop(&mut self) {
        println!("  Destructor: value was {}", self.value);
    }
}

fn main() {
    println!("=== Use-After-Free Safety in Rust ===\n");
    
    // Example 1: Rust prevents use-after-drop
    println!("Example 1: Ownership prevents use-after-free");
    {
        let ptr = Box::new(42);
        println!("  Allocated: *ptr = {}", ptr);
        drop(ptr);  // Explicitly free
        
        // COMPILE ERROR: Can't use after drop!
        // println!("  Value: {}", *ptr);  // ERROR: value used after move
    }
    println!("  Rust prevents use-after-free at compile time!");
    
    // Example 2: Can't return reference to local variable
    println!("\nExample 2: No dangling references");
    println!("  This would be a compile error:");
    println!("  fn bad() -> &i32 {{ ");
    println!("      let x = 42;");
    println!("      &x  // ERROR: returns reference to local variable");
    println!("  }}");
    
    // Example 3: Proper ownership transfer
    println!("\nExample 3: Safe ownership transfer");
    let s1 = String::from("Hello");
    println!("  s1 = {}", s1);
    
    let s2 = s1;  // Ownership moved
    println!("  s2 = {}", s2);
    
    // COMPILE ERROR: Can't use s1 after move!
    // println!("  s1 = {}", s1);  // ERROR: value used after move
    println!("  s1 is no longer accessible (moved to s2)");
    
    // Example 4: Borrowing prevents use-after-free
    println!("\nExample 4: Borrowing rules");
    let mut data = vec![1, 2, 3];
    
    {
        let r = &data[0];
        println!("  Reference to first element: {}", r);
        
        // COMPILE ERROR: Can't modify while borrowed!
        // data.push(4);  // ERROR: cannot borrow `data` as mutable
    }  // Borrow ends here
    
    data.push(4);  // Now safe to modify
    println!("  Added element after borrow ended");
    
    // Example 5: No iterator invalidation
    println!("\nExample 5: Iterator safety");
    let vec = vec![1, 2, 3, 4, 5];
    
    for val in &vec {
        println!("  Value: {}", val);
        // COMPILE ERROR: Can't modify vec while iterating!
        // vec.push(6);  // ERROR: cannot borrow as mutable
    }
    
    // Example 6: Rc for shared ownership
    println!("\nExample 6: Shared ownership with Rc");
    use std::rc::Rc;
    
    let data = Rc::new(42);
    println!("  Created Rc: {}", data);
    
    {
        let data2 = Rc::clone(&data);
        println!("  Cloned Rc: {}", data2);
        println!("  Strong count: {}", Rc::strong_count(&data));
    }  // data2 dropped, but data still valid
    
    println!("  Original still valid: {}", data);
    println!("  Strong count: {}", Rc::strong_count(&data));
    
    // Example 7: Objects with Drop trait
    println!("\nExample 7: RAII with Drop");
    {
        let obj = MyStruct::new(100);
        println!("  Using object: {}", obj.value);
    }  // Automatically dropped
    println!("  Object automatically cleaned up");
    
    // Example 8: Box prevents use-after-free
    println!("\nExample 8: Box ownership");
    let boxed = Box::new(MyStruct::new(200));
    println!("  Boxed value: {}", boxed.value);
    
    // Moving Box transfers ownership
    let boxed2 = boxed;
    println!("  Moved to boxed2: {}", boxed2.value);
    
    // COMPILE ERROR: Can't use boxed after move!
    // println!("{}", boxed.value);  // ERROR: value used after move
    
    println!("\n=== Key Takeaways ===");
    println!("1. Ownership system prevents use-after-free");
    println!("2. Can't return references to local variables");
    println!("3. Move semantics prevent using freed memory");
    println!("4. Borrowing rules prevent dangling references");
    println!("5. Iterator invalidation prevented by borrow checker");
    println!("6. Rc/Arc provide safe shared ownership");
    println!("7. Drop trait ensures cleanup");
    println!("\nResult: Use-after-free is prevented at COMPILE TIME!");
}
